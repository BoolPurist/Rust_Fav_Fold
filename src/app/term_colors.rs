use colored::*;

pub fn color_promt_msg(to_color: &str) -> String {
    to_color.blue().to_string()
}
pub fn color_error_msg(to_color: &str) -> String {
    to_color.red().to_string()
}
pub fn color_exists_msg(to_color: &str) -> String {
    to_color.green().to_string()
}
pub fn color_not_found(to_color: &str) -> String {
    to_color.red().to_string()
}
