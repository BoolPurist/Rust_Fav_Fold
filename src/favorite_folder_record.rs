use serde::{Deserialize, Serialize};
use std::path::Path;

use crate::trimmed_not_empty_text::{NonEmptyText, NotEmptyTextError};

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
/// Struct to encapsulate a valid favorite created by the user
/// This favorite is a not empty label with an absolute path
/// That path does need to point to existing location.
pub struct FavoriteFolderPath {
    name: NonEmptyText,
    location: NonEmptyText,
}

impl FavoriteFolderPath {
    /// # Errors
    ///
    /// - If name or path are empty or white spaces.
    /// - If path can not be resolved to a valid utf 8 string.
    /// - If a relative path can not be resolved to an absolute one.
    pub fn new(name: NonEmptyText, location: NonEmptyText) -> Self {
        Self { name, location }
    }

    pub fn set_name(&mut self, new_name: NonEmptyText) {
        self.name = new_name;
    }

    pub fn get_name(&self) -> &str {
        &self.name
    }

    pub fn path_str(&self) -> &str {
        &self.location
    }
    pub fn path(&self) -> &Path {
        Path::new(self.location.as_str())
    }
    pub fn does_exit(&self) -> bool {
        Path::new(self.location.as_str()).exists()
    }
}

/// Error to describe the cause of failing to create a struct [`FavoriteFolderPath`]
use thiserror::Error;
#[derive(Debug, Error)]
pub enum InvalidFavoriteFields {
    #[error("Name of favorite folder should not be empty or only whitespace")]
    EmptyName(#[from] NotEmptyTextError),
    #[error("Path of favorite folder should not be empty or only whitespace")]
    EmptyPath,
    #[error("Path of favorite folder is not a valid utf 8 text")]
    InvalidUtf8PathStr,
}
