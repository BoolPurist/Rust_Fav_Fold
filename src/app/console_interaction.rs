use super::term_colors;
use crate::{cli_args::GetParams, favorite_folder_record::FavoriteFolderPath, AppResult};
use std::io::Write;

pub fn ask_possible_prompt_for_ask_number(
    paths: &[FavoriteFolderPath],
    get_params: &GetParams,
) -> AppResult<Option<usize>> {
    if !get_params.copy_ask_number() {
        return Ok(None);
    }
    let len = paths.len();
    let prompt_message = format!("Enter a whole number between {} and {}: ", 1, len);
    let user_input = read_line_from_user(&prompt_message)?;

    let parsed_number: usize = user_input
        .trim()
        .parse()
        .map_err(|_| gen_req_number_message(len))?;

    if parsed_number == 0 || parsed_number > len {
        Err(gen_req_number_message(len).into())
    } else {
        Ok(Some(parsed_number))
    }
}

pub fn read_line_from_user(prompt: &str) -> AppResult<String> {
    let colored_prompt = term_colors::color_promt_msg(prompt);
    print!("{colored_prompt}");
    std::io::stdout().flush()?;

    let mut buffer = String::new();

    std::io::stdin().read_line(&mut buffer)?;

    Ok(buffer)
}

fn gen_req_number_message(max: usize) -> String {
    format!("Number is not between {} and {}", 1, max)
}
