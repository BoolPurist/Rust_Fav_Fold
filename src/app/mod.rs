use log::error;

use crate::favorite_folder_record::FavoriteFolderPath;
use crate::{cli_args::GetParams, favorite_table};
use std::error::Error;

use crate::{clipboard, file_access, AppResult};

pub mod console_interaction;
pub mod term_colors;

pub fn put_into_clipboard_or_print(content: &str, clipboard: bool) -> AppResult {
    if clipboard {
        clipboard::put_into_clipboard(content)?;
    } else {
        println!("{content}");
    }

    Ok(())
}

pub fn exit_with_error(message: &dyn Error) {
    let red_msg = term_colors::color_error_msg(&format!("Error: {message}"));
    error!("{}", red_msg);
    std::process::exit(1);
}

pub fn handle_get_subcommand(get_params: &GetParams) -> AppResult<String> {
    return match get_params.get_name() {
        Some(name_given) => {
            let favorites = file_access::get_favorites()?;
            let name = name_given.try_into()?;
            let content = match favorites.get(name) {
                Some(found) => found.path_str(),
                None => {
                    return match (get_params.copy_fuzzy(), get_params.copy_ask_number()) {
                        (false, false) => {
                            Err(format!("No path found for the name: {}", name_given).into())
                        }
                        (_, _) => get_all(&get_params),
                    };
                }
            };

            Ok(content.to_string())
        }
        None => get_all(&get_params),
    };

    fn get_all(get_params: &GetParams) -> AppResult<String> {
        let all_locations = file_access::get_favorites()?;

        let all_locations = if let Some(name) = get_params.get_name() {
            if get_params.copy_fuzzy() {
                let name = name.try_into()?;
                all_locations.filtered_containing_name(name)
            } else {
                all_locations
            }
        } else {
            all_locations
        };

        let content = draw_table_and_prompt(all_locations.as_slice(), get_params)?;

        Ok(content)
    }
}

fn draw_table_and_prompt(
    all_locations: &[FavoriteFolderPath],
    get_params: &GetParams,
) -> AppResult<String> {
    if all_locations.is_empty() {
        return Err(
            "No match found for given name or no labels were created so far"
                .to_string()
                .into(),
        );
    }

    let table = favorite_table::draw_favorite_table(all_locations, get_params.into());
    if get_params.copy_ask_number() {
        println!("{table}");

        let given_number =
            console_interaction::ask_possible_prompt_for_ask_number(all_locations, get_params)?;

        match given_number {
            Some(index_start_from_one) => {
                let index = index_start_from_one - 1;
                // function for asking number of user ensures that the index will not
                // be out of bounds
                let to_put = all_locations.get(index).unwrap().path_str().to_string();

                Ok(to_put)
            }
            None => Ok(table),
        }
    } else {
        Ok(table)
    }
}
