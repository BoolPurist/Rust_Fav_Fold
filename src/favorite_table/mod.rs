pub use draw_params::DrawParam;
mod draw_params;

use crate::app::term_colors;

use crate::favorite_folder_record::FavoriteFolderPath;
use shellexpand;
use std::borrow::Cow;
use std::path::Path;

#[derive(Debug)]
struct LineNameAndPath {
    name: String,
    path: String,
}

pub fn draw_favorite_table(all_locations: &[FavoriteFolderPath], params: DrawParam) -> String {
    let ask_numbers = params.ask_number();
    if params.clipboard() {
        draw_without_colors(all_locations, ask_numbers)
    } else {
        draw_with_colors(all_locations, ask_numbers)
    }
}

fn draw_without_colors(all_locations: &[FavoriteFolderPath], ask_numbers: bool) -> String {
    prepare_drawing(all_locations, ask_numbers)
        .into_iter()
        .map(|line| format!("{}{}", line.name, line.path))
        .collect::<Vec<String>>()
        .join("\n")
}

fn draw_with_colors(all_locations: &[FavoriteFolderPath], ask_numbers: bool) -> String {
    prepare_drawing(all_locations, ask_numbers)
        .into_iter()
        .map(|line| {
            let colored_path = get_colored_path_if_no_clipboard(&line.path);
            format!("{}{}", line.name, colored_path)
        })
        .collect::<Vec<String>>()
        .join("\n")
}

fn prepare_drawing(
    all_locations: &[FavoriteFolderPath],
    ask_numbers: bool,
) -> Vec<LineNameAndPath> {
    let spacing_padding = get_spacing_padding();

    let records_without_line_numbers =
        construct_recores_without_numbers(all_locations, &spacing_padding);
    let lines = if ask_numbers {
        handle_line_number_if_needed(
            all_locations,
            records_without_line_numbers,
            &spacing_padding,
        )
    } else {
        records_without_line_numbers.collect()
    };

    return lines;

    fn calc_max_width(all_locations: &[FavoriteFolderPath]) -> usize {
        all_locations
            .iter()
            .fold(0, |akk, next| akk.max(next.get_name().len()))
    }

    fn construct_recores_without_numbers<'a>(
        all_locations: &'a [FavoriteFolderPath],
        spacing_padding: &'a str,
    ) -> impl Iterator<Item = LineNameAndPath> + 'a {
        let max_width = calc_max_width(all_locations);
        all_locations.iter().map(move |next_record| {
            let name = next_record.get_name();
            assert!(
                name.len() <= max_width,
                "Could not find max width for all labels"
            );

            let padded_name = pad_from_right_until(name, max_width);
            let raw_path = next_record.path_str();

            let padded_name = format!("{padded_name}{spacing_padding}");
            LineNameAndPath {
                name: padded_name,
                path: raw_path.to_string(),
            }
        })
    }

    fn handle_line_number_if_needed(
        all_locations: &[FavoriteFolderPath],
        records_without_line_numbers: impl Iterator<Item = LineNameAndPath>,
        spacing_padding: &str,
    ) -> Vec<LineNameAndPath> {
        let max_number_width = all_locations.len().to_string().chars().count();

        records_without_line_numbers
            .enumerate()
            .map(move |(index, line)| {
                let number = index + 1;
                let to_pad = number.to_string();
                let padded_number = format!(
                    "{}{}{}",
                    pad_from_right_until(&to_pad, max_number_width),
                    spacing_padding,
                    line.name
                );

                LineNameAndPath {
                    name: padded_number,
                    path: line.path,
                }
            })
            .collect()
    }
}

fn get_colored_path_if_no_clipboard(raw_path: &str) -> String {
    let expanded = shellexpand::tilde(raw_path);
    let expanded_path = Path::new(expanded.as_ref());
    check_if_exits(expanded_path, raw_path)
}

fn check_if_exits(path: &Path, to_check_color: &str) -> String {
    if path.exists() {
        term_colors::color_exists_msg(to_check_color)
    } else {
        term_colors::color_not_found(to_check_color)
    }
}

/// Size of space between columns in a output.
/// Example: `Spacing` is 2 then between name and path is 2 whitespaces in the output for the
/// user.
const SPACING: usize = 2;

fn get_spacing_padding() -> String {
    " ".repeat(SPACING)
}

/// # Summary
///
/// Makes sure the return value is at least as long as `max_with` chars.
///
/// # Returns
///
/// Either returns the `to_pad` unchanged as ref or as new owned value with spaces appended from right
/// so the string is exactly `max_with` chars long.
fn pad_from_right_until(to_pad: &str, max_width: usize) -> Cow<'_, str> {
    let actual_len = to_pad.len();

    if actual_len >= max_width {
        return to_pad.into();
    }

    let diff = max_width - actual_len;
    let padding = " ".repeat(diff);
    let mut padded = String::from(to_pad);
    padded.push_str(&padding);

    padded.into()
}

#[cfg(test)]
mod testing {
    use crate::AllFavorites;

    use super::*;
    const INPUT: &str = include_str!("longer_test_input.json");

    fn given_initial() -> AllFavorites {
        let parsed: Vec<FavoriteFolderPath> = serde_json::from_str(INPUT).unwrap();
        AllFavorites::new(parsed)
    }

    #[test]
    fn list_all_every_favorite() {
        let given = given_initial();
        let actual = draw_without_colors(given.as_slice(), false);
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn list_all_with_prompt_every_favorite() {
        let given = given_initial();
        let actual = draw_without_colors(given.as_slice(), true);
        insta::assert_snapshot!(actual);
    }
}
