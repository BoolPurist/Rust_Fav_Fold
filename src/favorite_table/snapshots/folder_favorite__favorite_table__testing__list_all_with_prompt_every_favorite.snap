---
source: src/favorite_table/mod.rs
expression: actual
---
1  studium     /home/some_user/Documents/Studium
2  developing  /home/some_user/Code/pyting
3  looping     /home/some_user/looping
4  dev_rust    /home/some_user/Code/rust
5  proc_macro  /home/some_user/Code/rust/proc-macro-workshop
